src/storage.ts

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2021 Google Inc. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import { openDB, DBSchema, IDBPDatabase } from 'idb';
+
+const mainDBName = 'appDB';
+const mainImageName = 'mainImage';
+
+export type FileUniqueID = string;  // generated by the storage layer
+
+export type FileMetadata = {
+  filename: string;
+  mimeType: string;
+  title?: string;
+}
+
+export type FileRecord = {
+  id: FileUniqueID;
+  metadata: FileMetadata;
+  added: number;  // timestamp
+  blobModified: number;  // timestamp
+  metadataModified: number;  // timestamp
+};
+
+interface LlamaDB extends DBSchema {
+  blob: {
+    key: string;
+    value: Blob;
+  };
+  metadata: {
+    key: string;
+    value: FileRecord;
+  };
+}
+
+export default class LlamaStorage {
+  private db: IDBPDatabase<LlamaDB>;
+
+  private constructor(db: IDBPDatabase<LlamaDB>) {
+    console.log('Database initialized');
+    this.db = db;
+  }
+
+  static async create() {
+    return new LlamaStorage(await openDB(mainDBName, undefined /* version */, {
+      upgrade(db, oldVersion, newVersion, transaction) {
+        console.log(`DB update request`);
+        db.createObjectStore('blob');
+        db.createObjectStore('metadata');
+      },
+      // blocked()
+      // blocking()
+      // terminated()
+    }));
+  }
+
+  async add(file: Blob, metadata: FileMetadata): Promise<FileRecord> {
+    const id: FileUniqueID = mainImageName /* TODO: generate unique IDs */;
```

Comment:
It would probably be worth addressing this TODO in this PR (just having some incrementing int would be good enough) - at the moment FileUniqueId is very not-unique.

Reply:
this is a bit complicated. We want to support multiple images, but at the moment we really just don't. There's no solid plan for how the UI will work (in particular, how do we know which image to load immediately when you open the app? Is it in the URL? In another section of indexedDB?)

I think it needs a bit more discussion (https://github.com/GoogleChromeLabs/llaminator/issues/21 seems like a good place to discuss) before we're ready to unhardcode the main image ID.

Reply:
For what it's worth, I'm also tempted to say that this is a constraint on the presentation layer, not on the storage layer. Properly implementing this class to full functionality doesn't have to be blocked on solving UI concerns.

Reply:
(Doesn't have to be this PR, though.)